### PR TITLE
tests: Add fuzzing harness for CCoinsViewCache

### PR DESCRIPTION
**Pull request by practicalswift**
Add fuzzing harness for `CCoinsViewCache`.

See [`doc/fuzzing.md`](https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md) for information on how to fuzz Bitcoin Core. Don't forget to contribute any coverage increasing inputs you find to the [Bitcoin Core fuzzing corpus repo](https://github.com/bitcoin-core/qa-assets).

Happy fuzzing :)

### CODE REVIEW

File: src/test/fuzz/coins_view.cpp [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            if (random_coin.IsSpent()) {
+                break;
+            }
+            Coin coin = random_coin;
+            bool expected_code_path = false;
+            try {
+                coins_view_cache.AddCoin(random_out_point, std::move(coin), fuzzed_data_provider.ConsumeBool());
+                expected_code_path = true;
+            } catch (const std::logic_error& e) {
+                if (e.what() == std::string{"Attempted to overwrite an unspent coin (when possible_overwrite is false)"}) {
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:64]:
```suggestion
                if (e.what() != std::string{"Attempted to overwrite an unspent coin (when possible_overwrite is false)"}) raise e;
```

Could `expected_code_path` be removed with this fixup? If yes, that seems preferable, as it is less code.

Comment (practicalswift) [reply src/test/fuzz/coins_view.cpp:64]: The problem with doing that is that the `throw;` line (I assume `raise e;` was a typo :)) will not be covered which hinders the road to 100% line coverage :)

**Comment by jb55** [comment]
nice, Concept ACK

**Comment by laanwj** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/18867/commits/b5f971a16ecbfcb0487b274f09516f17dfa19ab6

**Comment by practicalswift** [comment]
@MarcoFalke Would you mind reviewing? :)

**Comment by practicalswift** [comment]
Rebased! :)

File: src/test/fuzz/coins_view.cpp [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    while (fuzzed_data_provider.ConsumeBool()) {
+        switch (fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 9)) {
+        case 0: {
+            if (random_coin.IsSpent()) {
+                break;
+            }
+            Coin coin = random_coin;
+            bool expected_code_path = false;
+            try {
+                coins_view_cache.AddCoin(random_out_point, std::move(coin), fuzzed_data_provider.ConsumeBool());
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:61]:
```suggestion
                coins_view_cache.AddCoin(random_out_point, std::move(coin), possible_overwrite);
```

File: src/test/fuzz/coins_view.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+                expected_code_path = true;
+            } catch (const std::logic_error& e) {
+                if (e.what() == std::string{"Attempted to overwrite an unspent coin (when possible_overwrite is false)"}) {
+                    expected_code_path = true;
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:65]:
```suggestion
                    assert(possible_overwrite);
```

Could make this check a bit stricter?

File: src/test/fuzz/coins_view.cpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                        break;
+                    }
+                    coins_cache_entry.coin = *opt_coin;
+                }
+                if (fuzzed_data_provider.ConsumeBool()) {
+                    const std::optional<COutPoint> opt_out_point = ConsumeDeserializable<COutPoint>(fuzzed_data_provider);
+                    if (!opt_out_point) {
+                        break;
+                    }
+                    random_out_point = *opt_out_point;
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:138]: What is this doing? How is this different from the fuzz engine running case 6 -> case 9?

File: src/test/fuzz/coins_view.cpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+                }
+                coins_map.emplace(random_out_point, std::move(coins_cache_entry));
+            }
+            bool expected_code_path = false;
+            try {
+                coins_view_cache.BatchWrite(coins_map, fuzzed_data_provider.ConsumeBool() ? ConsumeUInt256(fuzzed_data_provider) : coins_view_cache.GetBestBlock());
+                expected_code_path = true;
+            } catch (const std::logic_error& e) {
+                if (e.what() == std::string{"FRESH flag misapplied to coin that exists in parent cache"}) {
+                    expected_code_path = true;
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:148]: Note to myself only: I thought about putting a more strict assert here, but the exception doesn't give any hints which coin failed, so I don't think this is possible for now.

File: src/test/fuzz/coins_view.cpp [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+                // coins.cpp:69: void CCoinsViewCache::AddCoin(const COutPoint &, Coin &&, bool): Assertion `!coin.IsSpent()' failed.
+                break;
+            }
+            bool expected_code_path = false;
+            try {
+                AddCoins(coins_view_cache, transaction, fuzzed_data_provider.ConsumeIntegral<int>(), fuzzed_data_provider.ConsumeBool());
+                expected_code_path = true;
+            } catch (const std::logic_error& e) {
+                if (e.what() == std::string{"Attempted to overwrite an unspent coin (when possible_overwrite is false)"}) {
+                    expected_code_path = true;
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:227]: Same

File: src/test/fuzz/coins_view.cpp [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+            assert(expected_code_path);
+            break;
+        }
+        case 1: {
+            (void)AreInputsStandard(CTransaction{random_mutable_transaction}, coins_view_cache);
+            break;
+        }
+        case 2: {
+            TxValidationState state;
+            CAmount tx_fee;
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:239]:
```suggestion
            CAmount tx_fee_out;
```

File: src/test/fuzz/coins_view.cpp [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+            const CTransaction transaction{random_mutable_transaction};
+            if (ContainsSpentInput(transaction, coins_view_cache)) {
+                // Avoid:
+                // consensus/tx_verify.cpp:171: bool Consensus::CheckTxInputs(const CTransaction &, TxValidationState &, const CCoinsViewCache &, int, CAmount &): Assertion `!coin.IsSpent()' failed.
+                break;
+            }
+            try {
+                (void)Consensus::CheckTxInputs(transaction, state, coins_view_cache, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, std::numeric_limits<int>::max()), tx_fee);
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:247]: can `assert(MoneyRange(tx_fee_out));` when check was successful?

**Review by maflcko — APPROVED** [review]
ACK. Left some ideas for more `assert`s

68dd5ee2c0934b6cb8f595b4a97ba6e5d5e2a31a

**Comment by practicalswift** [comment]
@MarcoFalke Thanks a lot for a very good review! All feedback addressed :)

File: src/test/fuzz/coins_view.cpp [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    Coin random_coin;
+    CMutableTransaction random_mutable_transaction;
+    while (fuzzed_data_provider.ConsumeBool()) {
+        switch (fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 9)) {
+        case 0: {
+            if (random_coin.IsSpent()) {
+                break;
+            }
+            Coin coin = random_coin;
+            bool expected_code_path = false;
```
Comment (MarcoFalke) [inline src/test/fuzz/coins_view.cpp:59]: why is this needed?

Comment (MarcoFalke) [reply src/test/fuzz/coins_view.cpp:59]: Are you saying that the only exception that is allowed to be thrown is the one that is caught by the error string? Fair enough.

**Comment by MarcoFalke** [comment]
ACK f9b22e3bdb

**Comment by MarcoFalke** [comment]
ACK f9b22e3bdb 📫

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK f9b22e3bdb 📫
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhWqQv/XMjZcKGe3IffrT2bXwemMhi9fKPwO9pJcYGnW6+V/8iHWrTF44+knH5Z
k+gO6xywZRkw2LwHFQ2DRK8YYQex+AtAK4gsMu13bGZn6NoJToqgjfN4FZzZafzn
ZJmyOlPP2jO2GKAh3ur0ymUVRB5IvuitTnkjHJpFYgv0T+fFEEmrT+IwB6Gj+c9i
vjRTqU1Pg4eYxtS/zygb1faD6IjN2QUK/uiIV6wdy1aXBVxivx3/p7lCgN9J7M16
TrgvkdTX8Ox4mDMHV+bxmmDKPkDM2Z3ZmAlG2fr4l6ZQF57wQZ4InXRLJS/Rqx/Z
R3UzNufWomLHuwvuOyCTuoVuEjk6MmRmBYxa1mTD1JdC1zEdDBCGrB6Yc/TUzHKI
oUSIDxnvkOq4NGetBBuR0KwWuMOvpGfywV5zryVeMYvD4FtbrIauzDCF9Khjwqtf
Znhrmdf70sQDm8kU36/w64Iph54oERYYsV4DbhxqeGJQ+K1j6YKzOIvR7bPsS7ZX
k+eVRXEB
=x9vw
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `ead5a215f8a35eb59ebc41b959be5926bd52e9eb70983fd9272038e5d010f341  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108ead5a215f8a35eb59ebc41b959be5926bd52e9eb70983fd9272038e5d010f341f01060f3f2d4d14a0f92ca457cb561f2e18208fff01049422a8ddc924f15f2f0b4b0b6cde07208f1045eccf8daf008a3830298c0ee542c0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a50938d5623078cc8f38ecea5fcdad1f08f1045eccf8daf0082dc11e9d1ad5d5e00083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010aac5eab51ff79a51fb0e6828cad998a208f1045eccf8daf0084e7220ac0954ec620083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010df86f8e63ed97bba6e242d0a5c7d815108f1045eccf8daf0088090df4b7a8a98380083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/test/fuzz/coins_view.cpp (hunk 9 above)
Comment (practicalswift) [reply src/test/fuzz/coins_view.cpp:59]: Exactly :)
